### PR TITLE
Enhancements for ARMv8 crypto extensions specific to AES-CTR mode

### PR DESCRIPTION
**Pull request by b49020**
Enhancements:
- arm64: CE: AES-CTR: Add support for 64 bit IV
- arm64: CE: AES-CTR: update IV after partial final CTR block

### CODE REVIEW

File: core/lib/libtomcrypt/src/ciphers/aes/aes_modes_armv8a_ce_a64.S [hunk 1]
```diff
@@ -466,6 +467,7 @@ ENTRY(ce_aes_ctr_encrypt)
 	adds            x6, x6, #1              /* increment BE ctr */
 	rev             x7, x6
 	ins             v4.d[1], x7
+	cbnz            x9, .Lctrcarrydone      /* ignore overflow */
```
Comment (jenswi-linaro) [inline core/lib/libtomcrypt/src/ciphers/aes/aes_modes_armv8a_ce_a64.S]: ignore overflow for 64-bit counter value.

Comment (b49020) [reply core/lib/libtomcrypt/src/ciphers/aes/aes_modes_armv8a_ce_a64.S]: Yeah that's the intended behaviour. Do you suggest to update comment?

File: core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h [hunk 2]
```diff
@@ -466,11 +466,12 @@ extern const struct ltc_cipher_descriptor {
        @param ct      Ciphertext
        @param blocks  The number of complete blocks to process
        @param IV      The initial value (input/output)
+       @param IV_size The initial value size
```
Comment (jenswi-linaro) [inline core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h]: There's a comment in `aes_ctr_encrypt_nblocks()`:

> /* In big endian mode IV_size=0 -> 128 bits and IV_size=8 -> 64 bits */

How does that relate to this value, is the value in bytes/bits or on some other magic scale?

File: core/lib/libtomcrypt/src/ciphers/aes/aes_modes_armv8a_ce_a64.S (hunk 1 above)
Comment (jenswi-linaro) [reply core/lib/libtomcrypt/src/ciphers/aes/aes_modes_armv8a_ce_a64.S]: Yes please.

File: core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h (hunk 2 above)
Comment (b49020) [reply core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h]: Have a look at computation of CTR length for big endian case [here](https://github.com/OP-TEE/optee_os/blob/master/core/lib/libtomcrypt/src/modes/ctr/ctr_start.c#L48).

Basically for BE case CTR length is reversed meaning 0 -> 16 bytes, 1 -> 15 bytes, ..., 8 -> 8 bytes, and so on.

Comment (b49020) [reply core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h]: For LE case, it would be normal: 0 -> 0 bytes, 1 -> 1 bytes, and so on.

Comment (jenswi-linaro) [reply core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h]: My point is that the comment "The initial value size" is not very specific. You need to search the code to get some clue and even then it's not obvious.

Comment (b49020) [reply core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h]: How about "The initial value size identifier"?

**Comment by jforissier** [comment]
This PR should be open against the Libtomcrypt import branch, so that we can keep track of local changes and rebase more easily.

File: core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h (hunk 2 above)
Comment (jenswi-linaro) [reply core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h]: I think a complete description is needed.

**Comment by b49020** [comment]
@jforissier I tried to change base to `import/libtomcrypt-1.18.2-develop-20180819` branch but it seems like there are conflicts. Is this the correct branch?

File: core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h (hunk 2 above)
Comment (b49020) [reply core/lib/libtomcrypt/src/headers/tomcrypt_cipher.h]: Will add complete description.

**Comment by jforissier** [comment]
> @jforissier I tried to change base to `import/libtomcrypt-1.18.2-develop-20180819` branch but it seems like there are conflicts. Is this the correct branch?

Yes, it is. If `git rebase` gives conflicts, please try to apply your patch on top of this branch with `git cherry-pick`. It will likely work fine. You can open another PR and close this one if you like.

**Comment by b49020** [comment]
Ok will use `import/libtomcrypt-1.18.2-develop-20180819` branch to create PR.

**Comment by jforissier** [comment]
@b49020 great. BTW, I have a document describing this "import branch" thing: https://docs.google.com/document/d/1gnoIESWGKaBJw_s4tfQfOmvW1QXHR0Kk-2G1hirwJ4o/edit?usp=sharing
Perhaps it should be added to https://optee.readthedocs.io/.

**Comment by b49020** [comment]
Closing this PR as another PR: https://github.com/OP-TEE/optee_os/pull/3497 opened targeting `import/libtomcrypt-1.18.2-develop-20180819`.

**Pull request closed without merging**